src/native-desktop/RootView.tsx

```diff
@@ -0,0 +1,165 @@
+/**
+* RootView.tsx
+*
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT license.
+*
+* The top-most view that's used for proper layering or modals and popups.
+*/
+import React = require('react');
+import RN = require('react-native');
+import PropTypes = require('prop-types');
+
+import {RootView as RootViewBase, RootViewUsingProps as RootViewUsingPropsBase,
+        BaseRootViewProps, RootViewPropsWithMainViewType, RootViewState, BaseRootView} from '../native-common/RootView';
+import Input from './Input';
+import UserInterface from './UserInterface';
+import EventHelpers from '../native-common/utils/EventHelpers';
+import FocusManager from './utils/FocusManager';
+import FrontLayerViewManager from '../native-common/FrontLayerViewManager';
+
+type SyntheticEvent = React.SyntheticEvent<any>;
+
+const KEY_CODE_TAB = 9;
+const KEY_CODE_ESC = 27;
+
+const styles = RN.StyleSheet.create({
+    appWrapper: {
+      flex: 1,
+    },
+  });
+
+//
+// Mixin with keyboard management behaviors. It enhances the two RootView flavors by adding:
+// 1. Support for maintaining UserInterface.keyboardNavigationEvent
+//   React Native doesn't offer a convenient mechanism to peek into keyboard/mouse events globally (the way
+// addEventListener achieves in the Web counterpart), so we rely on a spying view we render.
+// 2. Redirection to Input.ts
+// 3. A place for the root FocusManager
+type Constructor<T extends React.Component> = new (...args: any[]) => T;
+
+function applyDesktopBehaviorMixin<TRootViewBase extends Constructor<React.Component>>(RootViewBase: TRootViewBase) {
+    return class RootView extends RootViewBase implements React.ChildContextProvider<any> {
+        static childContextTypes: React.ValidationMap<any> = {
+            focusManager: PropTypes.object
+        };
+
+        _focusManager: FocusManager;
+        _keyboardHandlerInstalled = false;
+        _isNavigatingWithKeyboard: boolean = false;
+        _isNavigatingWithKeyboardUpateTimer: number | undefined;
+
+        constructor(...args: any[]) {
+            super(...args);
+            // Initialize the root FocusManager which is aware of all
+            // focusable elements.
+            this._focusManager = new FocusManager(undefined);
+        }
+
+        _onTouchStartCapture = (e: SyntheticEvent) => {
```

Comment:
>_onTouchStartCapture [](start = 8, length = 20)

Does this get called for mouse input? If not, who calls `_updateKeyboardNavigationState` for mouse input?

Reply:
Yes. It's a "mouse down", basically.
One thing that I know RNW doesn't have it fully is the direct "on mouse move" event. It does have the "touch move", but that's something different (more like "mouse dragging"). I'll open an issue when I have a chance (and fix).

---
In reply to: [159357430](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357430) [](ancestors = 159357430)